src/ForSyDe/Shallow/MoC/CSDF.hs

```diff
@@ -0,0 +1,481 @@
+-- unzipCSDF Processes
+------------------------------------------------------------------------
+
+unzipCSDF :: [(Int, Int)] -> Signal ([a], [b]) -> (Signal a, Signal b)
+unzipCSDF [] _ = (NullS, NullS)
+unzipCSDF _ NullS = (NullS, NullS)
+unzipCSDF ((p1, p2) : ps) ((s1, s2) :- ss)
+  | length s1 /= p1 || length s2 /= p2 = error "unzipCSDF: Process does not produce correct number of tokens"
+  | otherwise = (signal s1 +-+ sr1, signal s2 +-+ sr2)
+  where (sr1, sr2) = unzipCSDF (ps ++ [(p1, p2)]) ss
+
+
+unzip3CSDF :: [(Int, Int, Int)] -> Signal ([a], [b], [c])
+           -> (Signal a, Signal b, Signal c)
+unzip3CSDF [] _ = (NullS, NullS, NullS)
+unzip3CSDF _ NullS = (NullS, NullS, NullS)
+unzip3CSDF ((p1, p2, p3) : ps) ((s1, s2, s3) :- ss)
+  | length s1 /= p1 || length s2 /= p2
+    || length s3 /= p3 = error "unzip3CSDF: Process does not produce correct number of tokens"
+  | otherwise = (signal s1 +-+ sr1, signal s2 +-+ sr2, signal s3 +-+ sr3)
+  where (sr1, sr2, sr3) = unzip3CSDF (ps ++ [(p1, p2, p3)]) ss
+
+
+unzip4CSDF :: [(Int, Int, Int, Int)] -> Signal ([a], [b], [c], [d])
+           -> (Signal a, Signal b, Signal c, Signal d)
+unzip4CSDF [] _ = (NullS, NullS, NullS, NullS)
+unzip4CSDF _ NullS = (NullS, NullS, NullS, NullS)
+unzip4CSDF ((p1, p2, p3, p4) : ps) ((s1, s2, s3, s4) :- ss)
+  | length s1 /= p1 || length s2 /= p2
+    || length s3 /= p3 || length s4 /= p4 = error "unzip4CSDF: Process does not produce correct number of tokens"
+  | otherwise = (signal s1 +-+ sr1, signal s2 +-+ sr2, signal s3 +-+ sr3, signal s4 +-+ sr4)
+  where (sr1, sr2, sr3, sr4) = unzip4CSDF (ps ++ [(p1, p2, p3, p4)]) ss
+
+
+------------------------------------------------------------------------
+--
+-- Helper functions (not exported!)
+--
+------------------------------------------------------------------------
+
+sufficient_tokens :: (Num a, Eq a, Ord a) => a -> Signal t -> Bool
+sufficient_tokens 0 _     = True
+sufficient_tokens _ NullS = False
+sufficient_tokens n (_:-xs)
+ = if n < 0 then
+     error "sufficient_tokens: n must not be negative"
+   else
+     sufficient_tokens (n-1) xs
+
+
+outputTokens :: [(a, b, c)] -> [b]
+outputTokens [] = []
+outputTokens ((_, b, _):xs) = b : outputTokens xs
+
+
+inpOut1n :: [(it, ot, [a] -> y)] -> [(it, Int, [a] -> [y])]
+inpOut1n [] = []
+inpOut1n ((it, _, f):xs) = (it, 1, \a -> [f a]) : inpOut1n xs
+
+inpOut2n :: [(it, ot, [a] -> [b] -> y)] -> [(it, Int, [a] -> [b] -> [y])]
+inpOut2n [] = []
+inpOut2n ((it, _, f):xs) = (it, 1, \a b -> [f a b]) : inpOut2n xs
+
+inpOut3n :: [(it, ot, [a] -> [b] -> [c] -> y)] -> [(it, Int, [a] -> [b] -> [c] -> [y])]
+inpOut3n [] = []
+inpOut3n ((it, _, f):xs) = (it, 1, \a b c -> [f a b c]) : inpOut3n xs
+
+inpOut4n :: [(it, ot, [a] -> [b] -> [c] -> [d] -> y)] -> [(it, Int, [a] -> [b] -> [c] -> [d] -> [y])]
+inpOut4n [] = []
+inpOut4n ((it, _, f):xs) = (it, 1, \a b c d -> [f a b c d]) : inpOut4n xs
+
+------------------------------------------------------------------------
```

Comment:
Take out from here onwards. You might want to include some tests as in-line code examples.